Add spec for team endpoints

se7entyse7en opened the pull request:
Signed-off-by: Lou Marvin Caraig <loumarvincaraig@gmail.com>

se7entyse7en commented:
I'll continue development starting from this and eventually fix according to issues you might raise.

se7entyse7en commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,183 @@ paths:
+        explode: false
+        in: path
+        name: id
+        # Not actually required, but cannot set optional path parameter:
```
@vmarkovtsev wdyt about this? Do you have a better approach?

vmarkovtsev replied:
You set it as `required: true` so we will not be able to call `PUT /team` to create a new one. 

I had a similar problem with `/become` and I solved it with a URL parameter. However, we've got a CRUD here...

So maybe I added `CREATE /reposet` because of this issue.

se7entyse7en replied:
Got it, it makes sense. I'll then add an endpoint similarly to `/reposet`

warenlg commented:
Wow, I'm curious to see how we label this PR...

warenlg commented:
Review Required, not bad!

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,183 @@ paths:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: There is an existing team with the same contents.
```
Yes, this is a very good check to avoid duplicate teams in our precious DB :+1:

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,90 @@ components:
       items:
         $ref: '#/components/schemas/DeveloperID'
       type: array
+    TeamMember:
```
I wonder if we can merge `TeamMember` with `User` which are reported in `/account`...

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,91 @@ components:
+      example:
+        login: github.com/gkwillie
+        id: github|60340680
+        native_id: 60340680
```
Let's not add `native_id` to the response, I cannot imagine why it would needed, and you'd better not know [how we get it](https://github.com/athenianco/athenian-api/blob/master/server/athenian/api/auth.py#L350) :smile:

se7entyse7en replied:
hahaha 😂 

Anyway, the only reason I put is to reuse `User` and it's a required property in `User`,

vmarkovtsev replied:
ok, then let's not touch it.

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,91 @@ components:
+      allOf:
+        - $ref: '#/components/schemas/User'
+        - properties:
+            login:
```
How about adding this to the `User` model? Thus we can delete this `allOf`.

se7entyse7en replied:
Indeed, probably that would the best thing. 👍

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,91 @@ components:
+          required:
+            - login
+    Team:
+      description: A list of members part of the same team.
```
```suggestion
      description: List of team members.
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,91 @@ components:
+          picture: https://avatars1.githubusercontent.com/u/5599208?v=4
+          updated: 2020-05-05T18:19:05Z
+      properties:
+        id:
```
```suggestion
        id:
          description: Team identifier.
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1260,91 @@ components:
+        id:
+          type: integer
+        name:
+          type: string
```
```suggestion
          description: Name of the team.
          type: string
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,183 @@ paths:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Team'
+          description: List of members in the specified team.
```
```suggestion
          description: List of the members in the specified team.
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,183 @@ paths:
+          description: The specified team was not found.
+      security:
+      - bearerAuth: []
+      summary: List a team's members. The user must be in the account that owns the
```
```suggestion
      summary: List the team's members. The user must belong to the account that owns the
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,183 @@ paths:
+    put:
+      operationId: update_team
+      parameters:
+      - description: Numeric identifier of the team to update, empty on create.
```
```suggestion
      - description: Numeric identifier of the team to update, empty on creation.
```

se7entyse7en commented:
@vmarkovtsev all issues should have been addressed, and I added the endpoint for team creation

se7entyse7en commented:
But I guess this cannot be merged until I do not generate the models and add the new `login` attribute in `User`.

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class Team(Model):
+    """NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
```
Can you please copy-paste the model description here

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+# coding: utf-8
```
```suggestion
```
(we don't need it in py3)

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+    Do not edit the class manually.
+    """
+
+    def __init__(self, name: str = None, members: List[User] = None):
```
```suggestion
    def __init__(self, name: Optional[str] = None, members: Optional[List[User]] = None):
```

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+        self._name = name
+        self._members = members
+
+    @classmethod
+    def from_dict(cls, dikt: dict) -> "Team":
+        """Returns the dict as a model
+
+        :param dikt: A dict.
+        :return: The Team of this Team.
+        """
+        return util.deserialize_model(dikt, cls)
+
```
```suggestion
```

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+        return util.deserialize_model(dikt, cls)
+
+    @property
+    def name(self):
```
```suggestion
    def name(self) -> str:
```

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+        Name of the team.
+
+        :return: The name of this Team.
+        :rtype: str
```
```suggestion
```

vmarkovtsev commented on server/athenian/api/models/web/team.py:
```diff
@@ -0,0 +1,86 @@
+        return self._name
+
+    @name.setter
+    def name(self, name):
```
```suggestion
    def name(self, name: str):
```
... and so on. The default generated code is not ideal.

vmarkovtsev commented on server/athenian/api/models/web/user.py:
```diff
@@ -131,7 +136,30 @@ def id(self, id: str):
 
     @property
-    def native_id(self) -> str:
+    def login(self):
```
```suggestion
    def login(self) -> str:
```

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1296,139 @@ components:
+      - account
+      - team
+      type: object
+    TeamID:
+      description: Team identifier.
+      example: Engineering
+      type: string
```
```suggestion
```
(not used)

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+            schema:
+              $ref: '#/components/schemas/TeamCreateRequest'
+        x-body-name: body
+      responses:
```
How about 422 Unprocessable Entity. We raise 400 on JSON syntax errors, so if we can find a more specific code for this error then it will be distinguishable.

dpordomingo replied:
`422` lgtm

se7entyse7en replied:
good point, I'd just go for 400 with different error message. I think that distinguishing these two different cases by different status code will be important when and if we're going to have clients different from the webapp.

se7entyse7en replied:
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/6ce2316202b0863ecbb1a3a98385b49771ffc446).

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+      x-openapi-router-controller: athenian.api.controllers.team_controller
+      x-codegen-request-body-name: body
+  /team/{id}:
+    delete:
```
We are likely to use ARRAY the same way as in reposets to migrate less tables, though IDK exactly how Marvin sees it.

dpordomingo replied:
A problem I see if we embed the users in the teams is that they can be outdated easily if we do not implement other update processes.
Also, if at some point we need to change their schema, the migration process would be also more complex.

se7entyse7en replied:
Yup, I'd rather user an ARRAY as we do for `reposets` 👍

dpordomingo replied:
are we embedding the whole user or only its id?

vmarkovtsev replied:
Only the ids - or, more precisely, their logins, if we allow out-of-account members

dpordomingo replied:
then lgtm.
(`login` can change, so it might be not the best way to identify the users in the embedded list)

se7entyse7en replied:
I still need to better see how this could be done, do you need the answer now? I don't think that the way it is going to implemented will change the spec.

dpordomingo replied:
I don't need an answer. I'm just providing info :D

se7entyse7en commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/CreatedIdentifier'
```
yup totally makes sense to me, I just almost blindly copied what was being sent during the `PUT`, gonna change it since all those info are available to the api 👍

se7entyse7en replied:
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/b197c83d65ad261309a3cdb0c546443eb8a8ae0c).

se7entyse7en commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: Concurrent requests that are finishing the account registration.
```
mhmmm tricky. tbh I added this to make as much similar to `/reposets/{id}`, but probably in this case even `200` with empty list would fit as well. I think that `409` in case of non-finished account registration fits for enpdoints involving write ops.

dpordomingo replied:
Agree on returning empty list for read ops, and error on write ones.

se7entyse7en commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+            schema:
+              $ref: '#/components/schemas/Team'
+        x-body-name: body
+      responses:
```
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/57e0f929faa220ecdf10e91a922d70325f407f27).

se7entyse7en commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,219 @@ paths:
+            application/json:
+              schema:
+                type: object
+          description: Empty object indicates a successful operation.
```
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/eebbedc6dcb31817057c34cddedde07b6cd6e1a8).

vmarkovtsev commented on server/athenian/api/models/web/user.py:
```diff
@@ -131,7 +136,32 @@ def id(self, id: str):
+        self._login = login
+
+    @property
+    def native_id(self):
```
```suggestion
    def native_id(self) -> str:
```

se7entyse7en commented:
Once approved I'll squash commits to have a sane history

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1083,6 +1308,152 @@ components:
+            picture: https://avatars1.githubusercontent.com/u/5599208?v=4
+            updated: 2020-05-05T18:19:05Z
+      properties:
+        id:
```
Should we move `id` inside the `Team`? All the places that I see have them together.

I am also worried about how the team listing will work. If we accept a list of github logins then we will not always have an `id` - that's an Auth0 user ID. We cannot make it optional for the `User` either. It looks like we have to return back our `TeamMember`.

se7entyse7en replied:
> Should we move id inside the Team? All the places that I see have them together.

Good point, indeed it was needed when the put needed to handle empty id. 👍 

> I am also worried about how the team listing will work. If we accept a list of github logins then we will not always have an id - that's an Auth0 user ID. We cannot make it optional for the User either. It looks like we have to return back our TeamMember.

Okay, I got the point. But this is related mostly to creation and update itself, isn't it? Instead of accepting a list of GH handles they should accept a list of ids.

vmarkovtsev replied:
IDK what's the product decision here: allow any valid github login, allow any github login from the account's orgs or allow registered account users...

se7entyse7en replied:
I'd go strict, that is users that are member of the account, that would simplify a lot. Also in that case no need to go back to `TeamMember` since all members would be regular `User`s. \cc @warenlg

vmarkovtsev commented on server/athenian/api/controllers/team_controller.py:
```diff
@@ -0,0 +1,48 @@
+    pass
+
+
+async def delete_team(request: AthenianWebRequest, body: dict) -> web.Response:
```
Not all the handlers have a `body` - DELETE, for example.

vmarkovtsev replied:
```suggestion
async def delete_team(request: AthenianWebRequest, id: int) -> web.Response:
```

dpordomingo commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,231 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/TeamUpdateRequest'
```
Is it a good idea to be able to change the `team.id` within the `TeamUpdateRequest`? Since the team to be updated is already identified by the called URL, the `team.id` could be omitted.

se7entyse7en replied:
good point, since it's already in the path itself 👍

vmarkovtsev replied:
This is a good point :+1:

se7entyse7en commented:
@vmarkovtsev in addition the schema I also need to update all the places dealing with `User` to include the `login` attribute as well.

se7entyse7en commented:
@vmarkovtsev @dpordomingo I updated the spec [here](https://github.com/athenianco/athenian-api/pull/334/commits/20fbb892bc072bc6865d9a08a6f82247686e634b):
- since as for Product requirement we need team  members that aren't users, a team now is not a list of `User`s, but a list of `TeamMember`s,
- for now, I reverted adding `login` to `User` since now it's not needed,
- added a `/contributors` endpoint to return all the contributors of an account.

@dpordomingo I didn't force-pushed, but it will be much easier to see the whole diff.

se7entyse7en commented:
@dpordomingo I just realized that from `/contributors` I need to return also avatars, name, etc. I'm on it.

se7entyse7en commented:
done [here](https://github.com/athenianco/athenian-api/pull/334/commits/b835cd1b88bef203239c907dec59f2a8488ff9ce).

se7entyse7en commented:
Anyway @dpordomingo I'll know better what kind of errors (non-200) will be returned once I'll implement the endpoints.

dpordomingo commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1082,6 +1082,12 @@ paths:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user is forbidden to list contributors belonging to the specified account.
```
I guess than when implementing this, all account members will be allowed to list account team members (since regular users are already [able to change release settings](https://athenianco.atlassian.net/browse/ENG-696?focusedCommentId=12352), I'd say that they should be also able to list team members) 
wdyt @warenlg ?

se7entyse7en replied:
yup as described in ENG-726 all users can manage teams.

vmarkovtsev commented on server/athenian/api/controllers/contributors_controller.py:
```diff
@@ -0,0 +1,10 @@
+
+
+async def get_contributors(request: web.Request, id: int) -> web.Response:
+    """List all the contributors belonging to the specified account.
```
We should really-really post this in a separate PR. This is becoming an unmanageable monster.

se7entyse7en replied:
Do you mind if we keep just the spec and the stub? The idea is that with this version of the spec everything needed for the settings page is ready. Then after I finished the implementation I'll move to change the filtering.

se7entyse7en replied:
This is actually also the first endpoint that I implemented that just needs to be opened 😄

vmarkovtsev commented on server/athenian/api/models/web/contributor.py:
```diff
@@ -0,0 +1,130 @@
+
+
+class Contributor(Model):
+    """Contributor details."""
```
If this is a spec description then we should be a bit more verbose here.
```suggestion
    """Details about a developer who contributed to some repositories owned by the account."""
```

se7entyse7en replied:
added both here and in spec

dpordomingo commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,278 @@ paths:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user does not belong to the specified account.
```
(not blocker) I'd say it should be returned a `403` if the user is not allowed to list the account contributors.

se7entyse7en replied:
Since errors are minor things to be completely defined now, I'd keep them as they're now and I'll review all of them on each endpoint that I'll implement to make the errors defined in the spec reflect the implementation.

vmarkovtsev commented on server/athenian/api/models/web/team_create_request.py:
```diff
@@ -0,0 +1,102 @@
+
+    def __init__(
+        self,
+            account: Optional[int] = None,
```
Indentation fault?

se7entyse7en replied:
yup 🤦

vmarkovtsev commented on server/athenian/api/models/web/team_update_request.py:
```diff
@@ -0,0 +1,98 @@
+    """Team update request."""
+
+    def __init__(
+            self,
```
I think we should de-indent -4 spaces here

vmarkovtsev commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -838,6 +838,278 @@ paths:
+# {% endif %}
+      x-openapi-router-controller: athenian.api.controllers.team_controller
+      x-codegen-request-body-name: body
+  /contributors/{id}:
+    get:
+      operationId: get_contributors
+      parameters:
+      - description: Numeric identifier of the account.
+        explode: false
+        in: path
+        name: id
+        required: true
+        schema:
+          type: integer
+        style: simple
+      responses:
+        200:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Contributors'
+          description: List of all contributors belonging to the specified account.
+        403:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/GenericError'
+          description: The user is forbidden to list contributors belonging to the specified account.
```
Hence this should go in a separate PR

se7entyse7en commented:
@vmarkovtsev for better review I'd rather open separate PRs for each endpoint after this.

dpordomingo commented on server/athenian/api/openapi/openapi.yaml:
```diff
@@ -1371,7 +1371,7 @@ components:
     Contributor:
       additionalProperties: false
-      description: Contributor details.
+      description: Details about a developer who contributed to some repositories owned by the account.
```
Can't be added new hires to their team till they make a contribution? Why not listing also organization members?

se7entyse7en replied:
It will behave the same way as `/filter/contributors` just without the time constraint. New hires and org members won't be shown in the filtering as well as now. Is this limitation ok @warenlg ?

warenlg replied:
Yes, it is ok, I don't want us to spend eng effort on new hires that have not contributed anywhere yet. The use case is too minor, and there is no need to handle it from our users.

vmarkovtsev commented on server/athenian/api/models/web/user.py:
```diff
@@ -15,6 +15,7 @@ class User(Model):
     def __init__(
         self,
         id: Optional[str] = None,
+        login: Optional[str] = None,
```
This must be covered in the tests + handled in `auth.py`. We can revert back if it is too much side work.

se7entyse7en replied:
Ah indeed! Sorry, as I said [here](https://github.com/athenianco/athenian-api/pull/334#issuecomment-627276698) I actually reverted it back from the spec, but then forgot to revert the model itself. I'll do when cleaning up the history since it's not required now and would require extra work.

vmarkovtsev replied:
:+1:

se7entyse7en commented:
I cleaned up the history, I'll merge this once CI gives the green light.

dpordomingo commented:
@marvin Are these new endpoints working at staging?
```
curl -X GET "https://api-staging.athenian.co/v1/teams/1" -H "accept: application/json"
```
No response, no headers...
Imo it should return an empty array if no teams were found.

Then I tried to create a team:
```
curl 'https://api-staging.athenian.co/v1/team/create' \
    --data-binary '{
        "account":1,
        "name":"Engineers",
        "members":["github.com/dpordomingo","github.com/se7entyse7en"]
    }'
```
No response, no headers...
It should return the ID of the created group.

Then I tried to list the teams again to receive the created one
```
curl -X GET "https://api-staging.athenian.co/v1/teams/1" -H "accept: application/json"
```
No response, no headers...

vmarkovtsev commented:
@dpordomingo No, they are not implemented yet.

se7entyse7en commented:
Indeed, this PR was meant to just prepare the spec and add the stubs as minimal as possible to just make tests pass and be able to merge the spec. The first endpoint that will be ready is `/contributors/{id}`: https://github.com/athenianco/athenian-api/pull/344
